File: .claude/skills/net-event-bus/examples/liveconfig.go
// Live config — the config service you no longer run (Go).
//
// One publisher and two subscribers, three in-process mesh nodes over loopback
// UDP. The publisher registers a channel, both subscribers join by name, and
// every config revision is pushed once and applied by each subscriber locally.
// There is no config server to poll, no cache to invalidate and no reload to
// coordinate.
//
// Run: go run liveconfig.go
//
// Expected final line: RESULT ok subscribers=2 applied=2 version=2
package main

import (
	"fmt"
	"log"
	"net"
	"strings"
	"time"

	mesh "github.com/ai-2070/net/go"
)

// 64 hex characters = 32 bytes. Every node in a mesh shares it.
var pskHex = strings.Repeat("42", 32)

const deliver = 5 * time.Second

func reserveAddr() string {
	conn, err := net.ListenPacket("udp", "127.0.0.1:0")
	if err != nil {
		log.Fatalf("reserve udp port: %v", err)
	}
	addr := conn.LocalAddr().String()
	_ = conn.Close()
	return addr
}

func build(seed byte) (*mesh.MeshNode, string) {
	addr := reserveAddr()
	node, err := mesh.NewMeshNode(mesh.MeshConfig{
		BindAddr:        addr,
		PskHex:          pskHex,
		IdentitySeedHex: strings.Repeat(fmt.Sprintf("%02x", seed), 32),
		HeartbeatMs:     200,
	})
	if err != nil {
		log.Fatalf("new mesh node: %v", err)
	}
	return node, addr
}

func handshake(responder, initiator *mesh.MeshNode, responderAddr string) {
	pub, err := responder.PublicKey()
	if err != nil {
		log.Fatalf("public key: %v", err)
	}
	done := make(chan error, 1)
	go func() {
		_, err := responder.Accept(initiator.NodeID())
		done <- err
	}()
	if err := initiator.Connect(responderAddr, pub, responder.NodeID()); err != nil {
		log.Fatalf("connect: %v", err)
	}
	if err := <-done; err != nil {
		log.Fatalf("accept: %v", err)
	}
}

// parse reads `v=<n>;mode=<name>`. Unreadable revisions are ignored.
func parse(payload []byte) (int, string, bool) {
	version := -1
	mode := ""
	for _, field := range strings.Split(string(payload), ";") {
		if rest, ok := strings.CutPrefix(field, "v="); ok {
			if _, err := fmt.Sscanf(rest, "%d", &version); err != nil {
				return 0, "", false
			}
		} else if rest, ok := strings.CutPrefix(field, "mode="); ok {
			mode = rest
		}
	}
	if version < 0 || mode == "" {
		return 0, "", false
	}
	return version, mode, true
}

// drain polls every shard. A published event lands on the shard derived from
// its stream id, so a consumer drains all of them — Go has no async iterator.
func drain(node *mesh.MeshNode, applied map[int]string) int {
	deadline := time.Now().Add(deliver)
	seen := 0
	for time.Now().Before(deadline) {
		quiet := true
		for shard := uint16(0); shard < 4; shard++ {
			events, err := node.RecvShard(shard, 64)
			if err != nil {
				continue
			}
			for _, event := range events {
				quiet = false
				seen++
				if version, mode, ok := parse(event.Payload); ok {
					applied[version] = mode
				}
			}
		}
		if quiet && seen > 0 {
			break
		}
		time.Sleep(20 * time.Millisecond)
	}
	return seen
}

func main() {
	publisher, addrPublisher := build(0xF1)
	one, _ := build(0xF2)
	two, _ := build(0xF3)

	handshake(publisher, one, addrPublisher)
	handshake(publisher, two, addrPublisher)

	for _, n := range []*mesh.MeshNode{publisher, one, two} {
		if err := n.Start(); err != nil {
			log.Fatalf("start: %v", err)
		}
	}
	defer func() {
		for _, n := range []*mesh.MeshNode{publisher, one, two} {
			_ = n.Shutdown()
		}
	}()

	// The publisher owns the channel config. No broker registers it.
	if err := publisher.RegisterChannel(mesh.ChannelConfig{
		Name:       "config/edge",
		Visibility: "global",
		Reliable:   true,
	}); err != nil {
		log.Fatalf("register channel: %v", err)
	}

	// Subscribers join by name; SubscribeChannel blocks on the publisher's ack,
	// so by the time it returns this node is in the roster.
	if err := one.SubscribeChannel(publisher.NodeID(), "config/edge"); err != nil {
		log.Fatalf("subscribe one: %v", err)
	}
	if err := two.SubscribeChannel(publisher.NodeID(), "config/edge"); err != nil {
		log.Fatalf("subscribe two: %v", err)
	}

	publish := func(version int, mode string) *mesh.PublishReport {
		report, err := publisher.Publish(
			"config/edge",
			[]byte(fmt.Sprintf("v=%d;mode=%s", version, mode)),
			mesh.PublishConfig{Reliability: "reliable"},
		)
		if err != nil {
			log.Fatalf("publish v%d: %v", version, err)
		}
		return report
	}

	first := publish(1, "blue")
	fmt.Printf("published v1 to %d of %d subscribers\n", first.Delivered, first.Attempted)

	second := publish(2, "green")
	fmt.Printf("published v2 to %d of %d subscribers\n", second.Delivered, second.Attempted)

	appliedOne := map[int]string{}
	appliedTwo := map[int]string{}
	drain(one, appliedOne)
	drain(two, appliedTwo)

	applied := 0
	for _, appliedMap := range []map[int]string{appliedOne, appliedTwo} {
		if _, ok := appliedMap[2]; ok {
			applied++
		}
	}
	fmt.Printf("subscriber one applied:    %v\n", appliedOne)
	fmt.Printf("subscriber two applied:    %v\n", appliedTwo)
	fmt.Printf("roster at publish time:    %d\n", second.Attempted)

	fmt.Printf("RESULT ok subscribers=%d applied=%d version=2\n", second.Attempted, applied)
}

File: .claude/skills/net-event-bus/examples/liveconfig.py
"""Live config — the config service you no longer run.

One publisher and two subscribers, three in-process mesh nodes over loopback
UDP. The publisher registers a channel, both subscribers join by name, and every
config revision is pushed once and applied by each subscriber locally. There is
no config server to poll, no cache to invalidate and no reload to coordinate.

The distributed-channel verbs live on the low-level binding (``net.NetMesh``),
not on ``net_sdk.MeshNode`` — the wrapper has no channel surface at all.

Run:

    python liveconfig.py

Expected final line: ``RESULT ok subscribers=2 applied=2 version=2``
"""

from __future__ import annotations

import threading
import time

from net import NetMesh

# 64 hex characters = 32 bytes. Every node in a mesh shares it.
PSK = "42" * 32

DELIVER_S = 5.0


def build(seed: int) -> NetMesh:
    return NetMesh(
        "127.0.0.1:0",
        PSK,
        identity_seed=bytes([seed]) * 32,
        heartbeat_interval_ms=200,
    )


def handshake(responder: NetMesh, initiator: NetMesh) -> None:
    """One side connects, the other accepts; both calls block."""
    errors: list[BaseException] = []

    def accept() -> None:
        try:
            responder.accept(initiator.node_id)
        except BaseException as error:  # noqa: BLE001 - surfaced below
            errors.append(error)

    thread = threading.Thread(target=accept, daemon=True)
    thread.start()
    time.sleep(0.05)
    initiator.connect(responder.local_addr, responder.public_key, responder.node_id)
    thread.join(timeout=5.0)
    if thread.is_alive():
        raise RuntimeError("handshake timed out")
    if errors:
        raise errors[0]


def parse(text: str) -> tuple[int, str] | None:
    """Parse ``v=<n>;mode=<name>``; unreadable revisions are ignored.

    ``StoredEvent.raw`` is the payload string itself, not a JSON envelope.
    """
    version = None
    mode = None
    for field in text.split(";"):
        if field.startswith("v="):
            try:
                version = int(field[2:])
            except ValueError:
                return None
        elif field.startswith("mode="):
            mode = field[5:]
    if version is None or mode is None:
        return None
    return (version, mode)


def drain(node: NetMesh, applied: dict[int, str]) -> int:
    """Drain every shard the bus could have routed a channel event to.

    Published events land on the shard derived from the stream id, so a consumer
    polls all of them — the low-level binding's ``poll_shard`` is the receive
    half that the ergonomic wrapper does not expose.
    """
    deadline = time.monotonic() + DELIVER_S
    seen = 0
    while time.monotonic() < deadline:
        quiet = True
        for shard in range(4):
            try:
                events = node.poll_shard(shard, 64)
            except Exception:  # noqa: BLE001 - an empty shard is not fatal
                continue
            for event in events:
                quiet = False
                seen += 1
                parsed = parse(event.raw)
                if parsed is not None:
                    applied[parsed[0]] = parsed[1]
        if quiet and seen > 0:
            break
        time.sleep(0.02)
    return seen


def main() -> None:
    publisher = build(0xF1)
    one = build(0xF2)
    two = build(0xF3)

    handshake(publisher, one)
    handshake(publisher, two)

    publisher.start()
    one.start()
    two.start()

    try:
        # The publisher owns the channel config. No broker registers it.
        publisher.register_channel("config/edge", visibility="global", reliable=True)

        # Subscribers join by name; subscribe_channel blocks on the publisher's
        # ack, so by the time it returns this node is in the roster.
        one.subscribe_channel(publisher.node_id, "config/edge")
        two.subscribe_channel(publisher.node_id, "config/edge")

        first = publisher.publish("config/edge", b"v=1;mode=blue")
        print(f"published v1 to {first['delivered']} of {first['attempted']} subscribers")

        second = publisher.publish("config/edge", b"v=2;mode=green")
        print(f"published v2 to {second['delivered']} of {second['attempted']} subscribers")

        applied_one: dict[int, str] = {}
        applied_two: dict[int, str] = {}
        drain(one, applied_one)
        drain(two, applied_two)

        applied = sum(1 for applied_map in (applied_one, applied_two) if 2 in applied_map)
        print(f"subscriber one applied:    {sorted(applied_one.items())}")
        print(f"subscriber two applied:    {sorted(applied_two.items())}")
        print(f"roster at publish time:    {second['attempted']}")

        print(
            f"RESULT ok subscribers={second['attempted']} applied={applied} version=2"
        )
    finally:
        publisher.shutdown()
        one.shutdown()
        two.shutdown()


if __name__ == "__main__":
    main()
